pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -0,0 +1,871 @@
+                        yield stream_ctx.part_end
+                        stream_ctx.part_end = None
+
+    async def _agent_event(
+        self,
+        stream_ctx: _RequestStreamContext,
+        agent_event: AgentStreamEvent,
+    ) -> AsyncGenerator[BaseEvent, None]:
+        """Handle an agent event and yield AG-UI protocol events.
+
+        Args:
+            stream_ctx: The request stream context to manage state.
+            agent_event: The agent event to process.
+
+        Yields:
+            AG-UI Server-Sent Events (SSE) based on the agent event.
+        """
+        self._logger.debug('agent_event: %s', agent_event)
+        if isinstance(agent_event, PartStartEvent):
+            if stream_ctx.part_end:
+                # End the previous part.
+                yield stream_ctx.part_end
+                stream_ctx.part_end = None
+
+            part: ModelResponsePart = agent_event.part
+            if isinstance(part, TextPart):
+                message_id: str = stream_ctx.new_message_id()
+                yield TextMessageStartEvent(
+                    type=EventType.TEXT_MESSAGE_START,
+                    message_id=message_id,
+                    role=Role.ASSISTANT.value,
+                )
+                stream_ctx.part_end = TextMessageEndEvent(
+                    type=EventType.TEXT_MESSAGE_END,
+                    message_id=message_id,
+                )
+                if part.content:
+                    yield TextMessageContentEvent(  # pragma: no cover
+                        type=EventType.TEXT_MESSAGE_CONTENT,
+                        message_id=message_id,
+                        delta=part.content,
+                    )
+            elif isinstance(part, ToolCallPart):  # pragma: no branch
+                stream_ctx.last_tool_call_id = part.tool_call_id
+                yield ToolCallStartEvent(
+                    type=EventType.TOOL_CALL_START,
+                    tool_call_id=part.tool_call_id,
+                    tool_call_name=part.tool_name,
+                )
+                stream_ctx.part_end = ToolCallEndEvent(
+                    type=EventType.TOOL_CALL_END,
+                    tool_call_id=part.tool_call_id,
+                )
+
+            elif isinstance(part, ThinkingPart):  # pragma: no branch
+                yield ThinkingTextMessageStartEvent(
+                    type=EventType.THINKING_TEXT_MESSAGE_START,
+                )
+                if part.content:  # pragma: no branch
+                    yield ThinkingTextMessageContentEvent(
+                        type=EventType.THINKING_TEXT_MESSAGE_CONTENT,
+                        delta=part.content,
+                    )
+                stream_ctx.part_end = ThinkingTextMessageEndEvent(
+                    type=EventType.THINKING_TEXT_MESSAGE_END,
+                )
+
+        elif isinstance(agent_event, PartDeltaEvent):
+            if isinstance(agent_event.delta, TextPartDelta):
```

Comment:
Same as up with `part`: let's move `agent_event.delta` into a variable as we're repeating it a lot